numpy/_core/src/multiarray/iterators.c

```diff
@@ -561,195 +489,154 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
-            goto fail;
+        ret = iter_subscript(self, ind);
+        Py_DECREF(ind);
+        goto finish;
+    }
+
+    // Single boolean index
+    else if (indices[0].type == HAS_0D_BOOL) {
+        /* Deprecated 2025-07, NumPy 2.4 */
+        if (DEPRECATE("Indexing flat iterators with a 0-dimensional boolean index is deprecated "
+                      "and may be removed in a future version. (Deprecated NumPy 2.4)") < 0) {
+            goto finish;
         }
-        if (istrue) {
-            return PyArray_ToScalar(self->dataptr, self->ao);
+        if (indices[0].value) {
+            ret = PyArray_ToScalar(self->dataptr, self->ao);
+            goto finish;
         }
         else { /* empty array */
             npy_intp ii = 0;
-            dtype = PyArray_DESCR(self->ao);
             Py_INCREF(dtype);
-            ret = (PyArrayObject *)PyArray_NewFromDescr(Py_TYPE(self->ao),
-                                     dtype,
-                                     1, &ii,
-                                     NULL, NULL, 0,
-                                     (PyObject *)self->ao);
-            return (PyObject *)ret;
+            ret = PyArray_NewFromDescr(Py_TYPE(self->ao),
+                                       dtype,
+                                       1, &ii,
+                                       NULL, NULL, 0,
+                                       (PyObject *)self->ao);
+            goto finish;
         }
     }
 
-    dtype = PyArray_DESCR(self->ao);
-    size = dtype->elsize;
+    PyArray_ITER_RESET(self);
 
-    /* set up a cast to handle item copying */
+    if (index_type == HAS_INTEGER) {
+        if (check_and_adjust_index(&indices[0].value, self->size, -1, NULL) < 0) {
+            goto finish;
+        }
+        PyArray_ITER_GOTO1D(self, indices[0].value);
+        ret = PyArray_ToScalar(self->dataptr, self->ao);
+        PyArray_ITER_RESET(self);
+        goto finish;
+    }
 
+    /* set up a cast to handle item copying */
     NPY_ARRAYMETHOD_FLAGS transfer_flags = 0;
     npy_intp one = 1;
+
     /* We can assume the newly allocated output array is aligned */
     int is_aligned = IsUintAligned(self->ao);
     if (PyArray_GetDTypeTransferFunction(
-                is_aligned, size, size, dtype, dtype, 0, &cast_info,
+                is_aligned, dtype_size, dtype_size, dtype, dtype, 0, &cast_info,
                 &transfer_flags) < 0) {
-        goto fail;
+        goto finish;
     }
 
-    /* Check for Integer or Slice */
-    if (PyLong_Check(ind) || PySlice_Check(ind)) {
-        start = parse_index_entry(ind, &step_size, &n_steps,
-                                  self->size, 0, 1);
-        if (start == -1) {
-            goto fail;
-        }
-        if (n_steps == ELLIPSIS_INDEX || n_steps == NEWAXIS_INDEX) {
-            PyErr_SetString(PyExc_IndexError,
-                            "cannot use Ellipsis or newaxes here");
-            goto fail;
+    if (index_type & HAS_SLICE) {
```

Comment:
```suggestion
    if (index_type == HAS_SLICE) {
```
But it doesn't matter, because it cannot have two kinds of indices.